src/date/expand-pattern.js

```diff
@@ -38,6 +39,128 @@ return function( options, cldr ) {
 		);
 	}
 
+	function getBestMatchPattern( path, skeleton ) {
+		var availableFormats, ratedFormats, format, pattern;
+
+		pattern = cldr.main([ path, skeleton ]);
+
+		if ( skeleton && !pattern ) {
+			availableFormats = cldr.main([ path ]);
+			ratedFormats = [];
+
+			for ( format in availableFormats ) {
+				ratedFormats.push({
+					format: format,
+					pattern: availableFormats[format],
+					rate: compareFormats( skeleton, format )
+				});
+			}
+
+			ratedFormats = ratedFormats
+				.filter( function( format ) {
+					return format.rate > -1;
+				} )
+				.sort( function( formatA, formatB ) {
+					return formatA.rate - formatB.rate;
+				});
+
+			if ( ratedFormats.length ) {
+				pattern = augmentFormat( skeleton, ratedFormats[0].pattern );
+			}
+		}
+
+		return pattern;
+	}
+
+	function repeatStr( str, count ) {
+		var i, result = "";
+		for ( i = 0; i < count; i++ ) {
+			result = result + str;
+		}
+		return result;
+	}
+
+	function normalizePatternType( char ) {
+		switch ( char ) {
+			case "e":
+			case "E":
+			case "c":
+				return "e";
+
+			case "M":
+			case "L":
+				return "L";
+
+			default:
+				return char;
+		}
+	}
+
+	function compareFormats( formatA, formatB ) {
+		var distance,
+			typeA,
+			typeB,
+			matchFound,
+			i,
+			j;
+
+		if ( formatA === formatB ) {
+			return 0;
+		}
+
+		formatA = formatA.match( datePatternRe );
+		formatB = formatB.match( datePatternRe );
+		if ( formatA.length === formatB.length ) {
+			distance = 1;
+			for ( i = 0; i < formatA.length; i++ ) {
+				typeA = normalizePatternType( formatA[i].charAt( 0 ) );
+				typeB = null;
+				matchFound = false;
+				for ( j = 0; j < formatB.length; j++ ) {
+					typeB = normalizePatternType( formatB[j].charAt( 0 ) );
+					if ( typeA === typeB ) {
+						break;
+					} else {
+						typeB = null;
+					}
+				}
+				if ( null === typeB ) {
+					return -1;
+				}
+				distance = distance + Math.abs( formatA[i].length - formatB[j].length );
+				if ( formatA[i].charAt( 0 ) !== formatB[j].charAt( 0 ) ) {
+					distance = distance + 1;
+				}
+			}
+			return distance;
+		}
+		return -1;
+	}
+
+	function augmentFormat( requestedSkeleton, bestMatchFormat ) {
+		var i, j, matchedType, matchedLength, requestedType, requestedLength;
+
+		requestedSkeleton = requestedSkeleton.match( datePatternRe );
+		bestMatchFormat = bestMatchFormat.match( datePatternRe );
+
+		for ( i = 0; i < bestMatchFormat.length; i++ ) {
+			matchedType = bestMatchFormat[i].charAt( 0 );
+			matchedLength = bestMatchFormat[i].length;
+			for ( j = 0; j < requestedSkeleton.length; j++ ) {
+				requestedType = requestedSkeleton[j].charAt( 0 );
+				requestedLength = requestedSkeleton[j].length;
+				if (
+					normalizePatternType( matchedType ) === normalizePatternType( requestedType ) &&
+					matchedLength < requestedLength
+				) {
+					bestMatchFormat[i] = repeatStr( matchedType, requestedLength );
+				}
+			}
+		}
+
+		return bestMatchFormat.join( "" );
+	}
```

Comment:
In general about the functions above, can `.match(datePatternRe)` and `normalizePatternType` be made in advance in the caller function so that you don't need to execute these over and over?